### PR TITLE
New hook actionPDFDeliverySlipGenerator for custom delivery pdf slips

### LINKED ISSUE #23884: Custom delivery pdf slips
#### Is your feature request related to a problem?

I need to be able to customize the delivery slips to apply the different formats and connect them to the different generators provided by the delivery platforms.

#### Describe the solution you'd like

Ideally, we should apply a hook so the code can by attached from the delivery modules or from specific modules.

#### Alternatives you've considered

At previous versions we could also apply some overrides to the core code to achieve this. This is not longer possible at the src classes. We could expand the overrides to the src classes.

#### Additional context

A similar hook is currently applied to invoices at https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/PDF/InvoicePdfGenerator.php#L60
We could also move that hook and generalize it to any PDF generation.

### PR DESCRIPTION
**Pull request by Guisardo**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop / 1.7.7.x
| Description?      | The hook actionPDFInvoiceRender should be replicated for the slip delivery pdf to allow the customizations.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/23884
| How to test?      | It requires a module that subscribes to the hook actionPDFDeliverySlipGenerator and apply changes to the PDF or connect the flow to external delivery systems.
| Possible impacts? | The hook list of the documentation should be updated also with this new hook.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23885)

### CODE REVIEW

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php [hunk 1]
```diff
@@ -71,6 +72,10 @@ public function generatePDF(array $orderId)
             throw new RuntimeException($this->translator->trans('The order cannot be found within your database.', [], 'Admin.Orderscustomers.Notification'));
         }
 
+        if (Hook::exec('actionPDFDeliverySlipGenerator', ['order' => $order])) {
+            return;
+        }
+
```
Comment (Progi1984) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php]:
```suggestion
```

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php [hunk 2]
```diff
@@ -71,6 +72,10 @@ public function generatePDF(array $orderId)
...
         $order_invoice_collection = $order->getInvoicesCollection();
 
         $pdf = new PDF($order_invoice_collection, PDF::TEMPLATE_DELIVERY_SLIP, Context::getContext()->smarty);
```
Comment (Progi1984) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php]:
```suggestion
        if (Hook::exec('actionPDFDeliverySlipRender', ['order_invoice_list' => $order_invoice_list])) {
            return;
        }

        $pdf = new PDF($order_invoice_collection, PDF::TEMPLATE_DELIVERY_SLIP, Context::getContext()->smarty);
```

I take the same hook as in `src/Adapter/PDF/OrderInvoicePdfGenerator.php` file.

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Great job. I have done a small improvement.

Could you add the hook to the SQL file and the `hook.xml` file ? Thanks

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php (hunk 2 above)
Comment (Guisardo) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Thanks for the feedback @Progi1984 , in fact, at the beginning, I started adding the hook like that, but short after, I realize that the hook for the delievery slip don't really need the invoices array. Invoice are directly related to the invoices pdf but not to the delivery one. This is why I end up passing the actual order object to the hook. What do you think?

**Comment by arouiadib** [comment]
> Great job. I have done a small improvement.
> 
> Could you add the hook to the SQL file and the `hook.xml` file ? Thanks

and to docs :)

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php (hunk 1 above)
Comment (Guisardo) [reply src/Adapter/PDF/DeliverySlipPdfGenerator.php]: Thanks for the feedback @Progi1984 , in fact, at the beginning, I started adding the hook like that, but short after, I realize that the hook for the delievery slip don't really need the invoices array. Invoice are directly related to the invoices pdf but not to the delivery one. This is why I end up passing the actual order object to the hook. What do you think?

**Comment by Guisardo** [comment]
> > Great job. I have done a small improvement.
> > Could you add the hook to the SQL file and the `hook.xml` file ? Thanks
> 
> and to docs :)

PR for the docs here: https://github.com/PrestaShop/docs/pull/998

File: install-dev/data/xml/hook.xml [hunk 3]
```diff
@@ -256,10 +256,20 @@
       <title>Order slip creation</title>
       <description>This hook is called when a new credit slip is added regarding client order</description>
     </hook>
-    <hook id="displayShoppingCartFooter">
-      <name>displayShoppingCartFooter</name>
-      <title>Shopping cart footer</title>
-      <description>This hook displays some specific information on the shopping cart's page</description>
```
Comment (Progi1984) [inline install-dev/data/xml/hook.xml:262]: Don't remove this hook

File: src/Adapter/PDF/DeliverySlipPdfGenerator.php [hunk 4]
```diff
@@ -71,9 +72,17 @@ public function generatePDF(array $orderId)
             throw new RuntimeException($this->translator->trans('The order cannot be found within your database.', [], 'Admin.Orderscustomers.Notification'));
         }
 
-        $order_invoice_collection = $order->getInvoicesCollection();
+        if (Hook::exec('actionPDFDeliverySlipGenerate', ['order' => $order])) {
```
Comment (Progi1984) [inline src/Adapter/PDF/DeliverySlipPdfGenerator.php:75]: Please use HookManager and add it in the constructor

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: install-dev/upgrade/sql/1.7.8.0.sql [hunk 5]
```diff
@@ -22,6 +22,8 @@ INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `po
   (NULL, 'displayAdminGridTableBefore', 'Display before Grid table', 'This hook adds new blocks before Grid component table.', '1'),
   (NULL, 'displayAdminGridTableAfter', 'Display after Grid table', 'This hook adds new blocks after Grid component table.', '1'),
   (NULL, 'displayAdminOrderCreateExtraButtons', 'Add buttons on the create order page dropdown', 'Add buttons on the create order page dropdown', '1'),
+  (NULL, 'actionPDFDeliverySlipGenerate', 'Delivery slip generation', 'This hook is executed before the common delivery slips is generated. If true is return by the handler, the system will interpreter that the slip is processed elsewhere and the common delivery slip will not be generated.', '1'),
```
Comment (PierreRambaud) [inline install-dev/upgrade/sql/1.7.8.0.sql:25]: You have to use the 1.7.9.0.sql instead

**Comment by matks** [comment]
Hello, it seems this PR is now stale.

I close this PR, following our [review guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/) "Pull Requests may be closed after 30 days of inactivity". If however you are still interested in it, please ping us :) we can reopen.

Thank you for your contribution.

**Pull request closed without merging**